A few ideas

ismirsehregal opened the pull request:
Hi,
Even though I haven’t had the time to create a data.table alternative I had another look at the code yesterday and made some changes:
-	Implemented query button
-	Implemented empty plot (if no data is returned)
-	Performance: Avoided repetitive db queries caused by load_filtered_data_from_db() (e.g. in renderPlotly).
-	Replaced checkboxGroupInput with selectInput
-	Using boxplot instead of lines+markers

Kind regards

ismirsehregal committed:
Add files via upload

-	Implemented Query Button
-	Performance: Avoided repetitive db queries by use of load_filtered_data_from_db() (e.g. in renderPlotly).
-	Replaced checkboxGroupInput with select selectInput

ismirsehregal committed:
Add files via upload

-	Implemented Query Button
-	Performance: Avoided repetitive db queries by use of load_filtered_data_from_db() (e.g. in renderPlotly).
-	Replaced checkboxGroupInput with select selectInput

ismirsehregal committed:
Delete server.R

ismirsehregal committed:
Delete ui.R

ismirsehregal committed:
Add files via upload

Box plot replaces scatter (better performance - less confusing)

silberzwiebel commented:
Many thanks for your contribution, very much appreciated!

I just had a quick look at it. Most of your changes are great. 
However, unfortunately, you also broke some functionality.
For the future, it might be easier if you open single pull request for single features (e.g., one for the submit button, another one for performance, another one for plot-type changes, etc.). Feel free to ask how to do this with git / github.
Possibly even more efficient would be if you join us in our weekly meetings at Cafe dreiklang :) 
Anyway, you may succeed with this PR as you like (address the following issues in this single PR or close this one and open several new ones).

These are the issues that would need to be resolved before I happily merge this PR:
- the "Zeitspannen" tab is not functional anymore because it has no query-button.
- maybe add (a second?) query button closer to the dropdown location menus?
- the hourly plot has no hover-text anymore
- potentially, the hourly plot needs a little explanation on what the boxes are showing (is it 95% of the data + outliers? is it mean + 3 SD + outliers?)? this tool is intended for people without much statistical knowledge
- while I generally like the new hourly plot, I'm not so sure about removing the lines in the yearly plot; it makes it harder to see "trends" in the data -- happy to discuss this further
- the selectInput is cool, didn't know about that. While it's easy to add items to the selection, I found it rather hard to know how to remove items. Is it possible to add little X icons to each item to remove them with a single click?
- please use a German UI (e.g., "anzeigen"/"abfragen" (?) instead of "query" for the button"; although we are writing here in English ;))

Thanks again for your contribution :+1:

ismirsehregal committed:
Add files via upload

- Fix query button
- German labels
- selectizeInput with remove-button

ismirsehregal committed:
Delete server.r

ismirsehregal committed:
Delete ui.r

ismirsehregal committed:
Add files via upload

- Fix query button
- German labels
- selectizeInput with remove-button

ismirsehregal commented:
Hi, thanks for the feedback! 
I already uploaded an updated version.

> _However, unfortunately, you also broke some functionality._

Oh, the button was hidden :)... should be working now.

>_open single pull request for single features_

Good to know, I'm new to GitHub. I'll try doing that next time.

>_Possibly even more efficient would be if you join us in our weekly meetings at Cafe dreiklang :)_

It's a 3h ride from my place :( .. but still it's fun following your project, talking to people using the same tools (R, shiny, plotly, docker) and learning from each other!

>_the "Zeitspannen" tab is not functional anymore because it has no query-button.
maybe add (a second?) query button closer to the dropdown location menus?_

This is solved in the update - let me know if the new location is okay.

>_the hourly plot has no hover-text anymore
potentially, the hourly plot needs a little explanation on what the boxes are showing (is it 95% of the data + outliers? is it mean + 3 SD + outliers?)? this tool is intended for people without much statistical knowledge_

The boxplots are based on the IQR: [see](https://help.plot.ly/what-is-a-box-plot/).
In my opinion boxplots are giving an intuitive visual impression of the distribution (even without knowing the IQR background), but for sure that is a matter of taste. Nevertheless, the currently displayed connection lines from hour 23 to 0 are misleading in my eyes. Single traces for every 0 to 23 hour sequence might be an alternative.

Installing the latest plotly version will introduce proper hover infos: [see](https://github.com/plotly/plotly.js/issues/2145)

```
library(devtools)
install_github("ropensci/plotly")
```

![image](https://user-images.githubusercontent.com/36849480/40344137-b76a7268-5d93-11e8-976b-10fe03cd8d78.png)

But that is something to manage in the docker file/container.

> _removing the lines in the yearly plot; it makes it harder to see "trends" in the data -- happy to discuss this further_

It wasn't intended to remove the lines here, mode is still "lines+markers". I don't know what's going wrong here - need to check this!

> _the selectInput is cool, didn't know about that. While it's easy to add items to the selection, I found it rather hard to know how to remove items. Is it possible to add little X icons to each item to remove them with a single click?_

Done in the update. I switched to selectizeInput for that.

> _please use a German UI (e.g., "anzeigen"/"abfragen" (?) instead of "query" for the button"; although we are writing here in English ;))_

That is also updated now.

BR

ismirsehregal committed:
Add files via upload

- fixed line plot for "years"-graph

ismirsehregal commented:
>> removing the lines in the yearly plot; it makes it harder to see "trends" in the data -- happy to discuss this further

>It wasn't intended to remove the lines here, mode is still "lines+markers". I don't know what's going wrong here - need to check this!

That's fixed now also. I needed to ungroup the data.frame before plotting, to get the lines displayed; [see](https://stackoverflow.com/questions/46102483/r-plotly-invisible-lines)

ismirsehregal committed:
Add files via upload

Aligned caption to other elements and made it eventReactive to the query button.
Showing legend also for a single trace.

silberzwiebel commented:
> It's a 3h ride from my place :( .. but still it's fun following your project, talking to people using the same tools (R, shiny, plotly, docker) and learning from each other!

:+1: Thanks even more for the contribution! Let me know if you need help to deploy this app using open data from your town 😇

> This is solved in the update - let me know if the new location is okay.

Possibly you could horizontally center the button? And, I had the idea to change the color of the button if a user changes any input. This should capture attention visually trigger that the user needs to press the button to change the plots. I tried to realize this idea [using this code](https://stackoverflow.com/a/43643162) but failed in my first attempt (I couldn't manage to ignore the content of the input-change and just react on *any* input change).
Feel free to open another pull request for this idea, if you want to implement it. No harm done, if you don't ;)

Regarding the boxplots: I saw some labels now on my machine, let's see what the server does.
I couldn't change the labels, though, and I saw something different than your screenshot:
![bildschirmfoto von 2018-05-26 23-19-30](https://user-images.githubusercontent.com/4533862/40580255-60790270-613b-11e8-8e0f-d9c3e943b09c.png)
It would be nice to customize the labels, if anyhow possible (potentially like: "Fahrrad, 8 Uhr, Median: 300, Min: 20, Max: 600").
Again, if you want to improve this (but first let's see whatever version of plotly is running in the docker image; I'd rather not use unstable version, though), feel free to open another pull request :)  If not, that's fine, too.

Anyway, I made small changes to your work:

- moved the legend back in the plots because otherwise they take too much horizontal space (I tried earlier to put them at the bottom but they overwrote the x-labels then. there is somewhere a plotly bug for this)
- moved the `ungroup` statement from the plot function to the data-aggregation-function
- some small code formatting

Then I merged the pull request and released a new version of the app. We are having some problems with our server right now but if they are solved your contribution should be online soon :)

silberzwiebel commented:
It's now online. Thanks again!

ismirsehregal commented:
I will have a look at your suggestions asap :)
